pennylane/qnn.py

```diff
@@ -0,0 +1,205 @@
+        weight_shapes (dict[str, tuple]): a dictionary mapping from all weights used in the QNode to
+            their corresponding sizes
+        output_dim (int): the dimension of data output from the QNode
+        weight_specs (dict[str, dict]): an optional dictionary for users to provide additional
+            specifications for weights used in the QNode, such as the method of parameter
+            initialization. This specification is provided as a dictionary with keys given by the
+            arguments of the `add_weight()
+            <https://www.tensorflow.org/api_docs/python/tf/keras/layers/Layer#add_weight>`__
+            (e.g., ``initializer``) method and values being the corresponding specification.
+        **kwargs: additional keyword arguments passed to the `Layer
+            <https://www.tensorflow.org/api_docs/python/tf/keras/layers/Layer>`__ base class
+    """
+
+    def __init__(
+        self,
+        qnode: QNode,
+        weight_shapes: dict,
+        output_dim: int,
+        weight_specs: Optional[dict] = None,
+        **kwargs
+    ):
+        if not CORRECT_TF_VERSION:
+            raise ImportError("KerasLayer requires TensorFlow version 2 and above")
+
+        self.sig = qnode.func.sig
+
+        if INPUT_ARG not in self.sig:
+            raise TypeError(
+                "QNode must include an argument with name {} for inputting data".format(INPUT_ARG)
+            )
+        if INPUT_ARG in set(weight_shapes.keys()):
+            raise ValueError(
+                "{} argument should not have its dimension specified in "
+                "weight_shapes".format(INPUT_ARG)
+            )
+        if set(weight_shapes.keys()) | {INPUT_ARG} != set(self.sig.keys()):
+            raise ValueError("Must specify a shape for every non-input parameter in the QNode")
+        if qnode.func.var_pos:
+            raise TypeError("Cannot have a variable number of positional arguments")
+        if qnode.func.var_keyword:
+            raise TypeError("Cannot have a variable number of keyword arguments")
+
+        self.qnode = to_tf(qnode, dtype=tf.keras.backend.floatx())
+        self.weight_shapes = {
+            weight: (tuple(size) if isinstance(size, Iterable) else (size,))
+            for weight, size in weight_shapes.items()
+        }
+        self.output_dim = output_dim[0] if isinstance(output_dim, Iterable) else output_dim
+
+        defaults = {
+            name for name, sig in self.sig.items() if sig.par.default != inspect.Parameter.empty
+        }
+        self.input_is_default = INPUT_ARG in defaults
+        if defaults - {INPUT_ARG} != set():
+            raise TypeError("Only the argument {} is permitted to have a default".format(INPUT_ARG))
+
+        self.weight_specs = weight_specs if weight_specs is not None else {}
+
+        self.qnode_weights = {}
+
+        super(KerasLayer, self).__init__(dynamic=True, **kwargs)
```

Comment:
what happens when `dynamic=False`? Any interaction with mutable/immutable QNodes?

Reply:
I haven't seen this kind of usage for `super` before. Can you explain why it is coded this way? :thinking:

Reply:
> I haven't seen this kind of usage for super before. Can you explain why it is coded this way? 🤔

I think this is the old Python 2.7 way of using `super()`, in Python 3 it should look like this:
```suggestion
        super().__init__(dynamic=True, **kwargs)
```

@co9olguy: `dynamic=True` indicates to Keras to keep the layer in eager mode. If you instead use `dynamic=False` (the default), Keras will attempt to autocompile the layer into a graph (using `tf.function`). However, graph mode is not supported by QNodes.